clang/lib/Format/NumericLiteralCaseFixer.cpp

```diff
@@ -0,0 +1,347 @@
+      Suffix(getTransform(CaseStyle.UpperCaseSuffix)) {
+  switch (Language) {
+  case FormatStyle::LK_CSharp:
+  case FormatStyle::LK_Java:
+  case FormatStyle::LK_JavaScript:
+    Separator = '_';
+    break;
+  case FormatStyle::LK_C:
+  case FormatStyle::LK_Cpp:
+  case FormatStyle::LK_ObjC:
+  default:
+    Separator = '\'';
+  }
+}
+
+QuickNumericalConstantParser::QuickNumericalConstantParser(
+    const StringRef &IntegerLiteral, const FormatParameters &Transforms)
+    : IntegerLiteral(IntegerLiteral), Transforms(Transforms),
+      Formatted(IntegerLiteral), PrefixBegin(Formatted.begin()),
+      PrefixEnd(Formatted.begin()), HexDigitBegin(Formatted.begin()),
+      HexDigitEnd(Formatted.begin()),
+      FloatExponentSeparatorBegin(Formatted.begin()),
+      FloatExponentSeparatorEnd(Formatted.begin()),
+      SuffixBegin(Formatted.begin()), SuffixEnd(Formatted.begin()) {}
+
+void QuickNumericalConstantParser::parse() {
+  auto Cur = Formatted.begin();
+  const auto End = Formatted.end();
+
+  bool IsHex = false;
+  bool IsFloat = false;
+
+  // Find the range that contains the prefix.
+  PrefixBegin = Cur;
+  if (Cur != End && *Cur == '0') {
+    ++Cur;
+    if (Cur != End) {
+      const char C = *Cur;
+      switch (C) {
+      case 'x':
+      case 'X':
+        IsHex = true;
+        ++Cur;
+        break;
+      case 'b':
+      case 'B':
+        ++Cur;
+        break;
+      case 'o':
+      case 'O':
+        // Javascript uses 0o as octal prefix.
+        ++Cur;
+        break;
+      default:
+        break;
+      }
+    }
+  }
+  PrefixEnd = Cur;
+
+  // Find the range that contains hex digits.
+  HexDigitBegin = Cur;
+  if (IsHex) {
+    Cur = std::find_if_not(Cur, End, [this, &IsFloat](char C) {
+      if (C == '.') {
+        IsFloat = true;
+        return true;
+      }
+      return C == Transforms.Separator || llvm::isHexDigit(C);
+    });
+  }
+  HexDigitEnd = Cur;
+
+  // Find the range that contains a floating point exponent separator.
+  // Hex digits have already been scanned through the decimal point.
+  // Decimal/octal/binary literals must fast forward through the decimal first.
+  if (!IsHex) {
+    Cur = std::find_if_not(Cur, End, [this, &IsFloat](char C) {
+      if (C == '.') {
+        IsFloat = true;
+        return true;
+      }
+      return C == Transforms.Separator || llvm::isDigit(C);
+    });
+  }
+  // The next character of a floating point literal will either be the
+  // separator, or the start of a suffix.
+  FloatExponentSeparatorBegin = Cur;
+  if (IsFloat) {
+    const char LSep = IsHex ? 'p' : 'e';
+    const char USep = IsHex ? 'P' : 'E';
+    Cur = std::find_if_not(
+        Cur, End, [LSep, USep](char C) { return C == LSep || C == USep; });
+  }
+  FloatExponentSeparatorEnd = Cur;
+
+  // Fast forward through the exponent part of a floating point literal.
+  if (!IsFloat) {
+  } else if (FloatExponentSeparatorBegin == FloatExponentSeparatorEnd) {
+  } else {
```

Comment:
```suggestion
  if (IsFloat && FloatExponentSeparatorBegin != FloatExponentSeparatorEnd) {
```
Empty `if`s are a bad style (in my opinion).